pkg/autopilot/controller/updates/periodicupdater.go

```diff
@@ -0,0 +1,182 @@
+
+	return &periodicUpdater{
+		ctx:               ctx,
+		log:               logrus.WithField("component", "periodic-updater"),
+		updateConfig:      updateConfig,
+		k8sClient:         k8sClient,
+		clusterID:         clusterID,
+		currentK0sVersion: currentK0sVersion,
+		apClientFactory:   apClientFactory,
+	}, nil
+}
+
+func (u *periodicUpdater) Config() *apv1beta2.UpdateConfig {
+	return &u.updateConfig
+}
+
+func (u *periodicUpdater) Run() error {
+	u.log.Debug("starting periodic updater")
+	checkDuration := time.Duration(10 * time.Minute)
+	// ENV var used only for testing purposes
+	if e := os.Getenv("K0S_UPDATE_PERIOD"); e != "" {
+		cd, err := time.ParseDuration(e)
+		if err != nil {
+			u.log.Errorf("failed to parse %s as duration for update checks: %s", e, err.Error())
+		} else {
+			checkDuration = cd
+		}
+	}
+	u.log.Debugf("using %s for update check period", checkDuration.String())
+	go func() {
+		// Check for update every checkDuration, return when context is cancelled
+		ticker := time.NewTicker(checkDuration)
+		u.ticker = ticker
+		defer ticker.Stop()
+		for {
+			select {
+			case <-u.ctx.Done():
+				u.log.Infof("parent context done, stopping polling")
+				return
+			case <-ticker.C:
+				u.checkForUpdate()
+			}
+		}
+	}()
+
+	return nil
+}
+
+func (u *periodicUpdater) Stop() {
+	// u.cancel()
+	if u.ticker != nil {
+		u.ticker.Stop()
+	}
+}
+
+func (u *periodicUpdater) checkForUpdate() {
+	u.log.Debug("checking for updates")
+	ctx, cancel := context.WithTimeout(u.ctx, 2*time.Minute)
+	defer cancel()
+
+	// Check if there's a token configured
+	var token string
+	tokenSecret := &corev1.Secret{}
+	if err := u.k8sClient.Get(ctx, crcli.ObjectKey{Name: "update-server-token", Namespace: "kube-system"}, tokenSecret); err != nil {
+		u.log.Infof("unable to get update server token: %v", err)
+	} else {
+		token = string(tokenSecret.Data["token"])
+	}
+
+	// Fetch the latest version from the update server
+	channelClient, err := uc.NewChannelClient(u.updateConfig.Spec.UpdateServer, u.updateConfig.Spec.Channel, token)
+	if err != nil {
+		u.log.Errorf("failed to create channel client: %v", err)
+		return
+	}
+
+	k8sClient, err := u.apClientFactory.GetClient()
+	if err != nil {
+		u.log.Errorf("failed to create k8s client: %v", err)
+		return
+	}
+	// Collect cluster info
+	ci, err := CollectData(ctx, k8sClient)
+	if err != nil {
+		u.log.Errorf("failed to collect cluster info: %s", err.Error())
+		return
+	}
+	extraHeaders := ci.AsMap()
+
+	latestVersion, err := channelClient.GetLatest(extraHeaders)
+	if err != nil {
+		u.log.Errorf("failed to get latest version: %v", err)
+		return
+	}
+	u.log.Debugf("got new version: %s", latestVersion.Version)
+	// Check if the latest version is newer than the current version
+	current := version.MustParse(u.currentK0sVersion)
+	new := version.MustParse(latestVersion.Version)
```

Comment:
Better not use `MustParse` inside functions because it `panic`s. It's ok for tests and package level `var foo = version.MustParse` or inside `func init()`. 

version 0.4.1 has the constraints and `CheckString(string) bool`